### PR TITLE
Implement `ignore()`

### PR DESCRIPTION
**Pull request by wlandau**
# Summary

In commands in the workflow plan, you can ignore(`some_code`) to

1. Force `drake` to not track dependencies in `some_code`, and
2. Ignore any changes in `some_code` when it comes to deciding which target are out of date.

# Help requested

`Drake` uses `standardize_command()` to create a reproducibly-tracked representation (kernel) of a command. Changes to the kernel invalidate the target. As part of this PR, code inside `ignore()` is removed from the kernel. I want to use proper code analysis rather than text analysis to to accomplish this, but I do not know how. The failing tests expose problems with the text analysis approach.

# GitHub issues fixed

- Ref: #299 

# Checklist

- [x] I have read `drake`'s [code of conduct](https://github.com/ropensci/drake/blob/master/CONDUCT.md), and I agree to follow its rules.
- [x] I have read the [guidelines for contributing](https://github.com/ropensci/drake/blob/master/CONTRIBUTING.md).
- [x] I have listed any substantial changes in the [development news](https://github.com/ropensci/drake/blob/master/NEWS.md).
- [x] I have added [`testthat`](https://github.com/r-lib/testthat) unit tests to [`tests/testthat`](https://github.com/ropensci/drake/tree/master/tests/testthat) to test new functionality.
- [x] Tests pass locally locally with `devtools::check()`
- [x] This pull request is ready for review.
- [x] Ready to merge.

### CODE REVIEW

File: R/workplan.R [hunk 1]
```diff
@@ -260,7 +260,7 @@ file_in <- function(...){
 #'   one file output per command. See the examples
 #'   for a full explanation.
 #' @export
-#' @seealso file_in knitr_in
+#' @seealso [file_in()], [knitr_in(()], [ignore()]
```
Comment (krlmlr) [inline R/workplan.R]: Two opening parens instead of one.

Comment (wlandau) [reply R/workplan.R]: Thanks: 379e5c7c377e7f6538d0073e89b7a389b7b5c5a2

**Comment by wlandau** [comment]
> I think standardize_command() could be implemented much in a same way like code_dependencies(): Walk the AST, return the input, unless it's an ignore() call.

I would like to, but if we [return the input](https://github.com/ropensci/drake/blob/e78d2c270bde1dd3a9d238fcc523ffd0ce1b6c69/R/dependencies.R#L338), would we risk getting a nested list instead of an expression or language object? I think that's the final piece of this PR/issue.

**Comment by krlmlr** [comment]
For walking and updating the AST and updating, assigning via `[<-` seems to preserve type:

``` r
cl_orig <- call("fun", quote(arg1), quote(arg2))
cl <- cl_orig
cl
#> fun(arg1, arg2)
cl[] <- cl[]
cl
#> fun(arg1, arg2)
identical(cl_orig, cl)
#> [1] TRUE
```

> Created on 2018-03-11 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by wlandau** [comment]
I'm not sure how `cl[] <- cl[]` will help us recurse on the elements of a language object, so I [put in a `for` loop](https://github.com/ropensci/drake/blob/59baef7504d3f077ee59a879fcccf295709b14d9/R/commands.R#L100) for now. Willing to discuss and improve. At least now we're using proper code analysis to walk the AST. And now `ignore()` appears to work not only for commands but also for imported functions.

File: R/commands.R [hunk 2]
```diff
@@ -75,18 +76,47 @@ standardize_command <- function(x) {
...
+    body(x) <- recurse_ignore(body(x))
+  } else if (
+    length(x) > 0 &&
+    is.language(x) &&
+    (is.call(x) || is.recursive(x))
+  ){
+    if (is_ignore_call(x)) {
+      x <- quote(ignore())
+    } else {
+      for (i in seq_along(x)){
```
Comment (krlmlr) [inline R/commands.R]: `x[] <- map(x, recurse_ignore)` ?

File: R/dependencies.R [hunk 3]
```diff
@@ -317,7 +317,7 @@ code_dependencies <- function(expr){
       new_globals <- setdiff(
         x = wide_deparse(expr), y = drake_fn_patterns)
       results$globals <<- c(results$globals, new_globals)
-    } else if (is.call(expr) || is.recursive(expr)) {
+    } else if (is.language(expr) && (is.call(expr) || is.recursive(expr))) {
```
Comment (krlmlr) [inline R/dependencies.R]: Can we extract this check to a function, maybe `is_callish()` ? Why do we need to check `is.recursive()` ?

File: R/dependencies.R [hunk 4]
```diff
@@ -459,3 +461,11 @@ is_loadd_call <- function(expr){
 is_readd_call <- function(expr){
   wide_deparse(expr[[1]]) %in% readd_fns
 }
+
+is_ignore_call <- function(expr){
+  wide_deparse(expr[[1]]) %in% ignore_fns
+}
+
+ignore_pattern <- paste0(
```
Comment (krlmlr) [inline R/dependencies.R]: Do we still need this?

File: R/commands.R (hunk 2 above)
Comment (wlandau) [reply R/commands.R]: `purrr::map(x, recurse_ignore)` and purrr::map(x[], recurse_ignore)` both give me "Error: `.x` is not a vector (language)".

File: R/dependencies.R (hunk 4 above)
Comment (wlandau) [reply R/dependencies.R]: Nope. Removed in b8d263693e0bddb80a06fa479e448a6e235ee856.

File: R/dependencies.R (hunk 3 above)
Comment (wlandau) [reply R/dependencies.R]: `is_callish()`, yes. See b8d263693e0bddb80a06fa479e448a6e235ee856
. For both `code_dependencies()` and `ignore_ignore()`, I remember thinking we don't need `is.recursive()` at first, but it in my experience it turns out to be needed for some edge cases: for example, diving into explicit `{` blocks.

File: R/commands.R (hunk 2 above)
Comment (krlmlr) [reply R/commands.R]: `map(as.list(x))` maybe?

Comment (wlandau) [reply R/commands.R]: Maybe, but then it seems like we would be changing the type again.

Comment (krlmlr) [reply R/commands.R]: Also with `x[] <- map(as.list(x), ...)` ?

``` r
x <- call("fun", quote(a1), quote(a2))
x
#> fun(a1, a2)
x[] <- purrr::map(as.list(x), ~as.name(paste0("x_", .)))
x
#> x_fun(x_a1, x_a2)
```

> Created on 2018-03-16 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

Comment (wlandau) [reply R/commands.R]: `x[] <- purrr::map(as.list(x), recurse_ignore)` works. Thanks, Kirill! Fixed in 3775026f1c899769a9bbcd9014f294cefe7595b5.

**Pull request merged**